### PR TITLE
rpc: add new client, use it everywhere

### PR DESCRIPTION
**Pull request by fjl**
The new client implementation supports concurrent requests,
subscriptions and replaces the various ad hoc RPC clients
throughout go-ethereum.

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Tue Jul 12 21:20:39 UTC 2016_

**Comment by fjl** [comment]
Please review: @bas-vk @karalabe

File: rpc/client.go [hunk 1]
```diff
@@ -0,0 +1,737 @@
...
+// subscription. Server notifications for the subscription are sent to the given channel.
+// The element type of the channel must match the expected type of content returned by the
+// subscription.
+
+// Callers should not use the same channel for multiple calls to EthSubscribe. The channel
+// is closed when the notification is unsubscribed or an error occurs. The error can be
+// retrieved via the Err method of the subscription.
+//
+// Slow subscribers will block the clients ingress path eventually.
+func (c *Client) EthSubscribe(channel interface{}, args ...interface{}) (*ClientSubscription, error) {
```
Comment (bas-vk) [inline rpc/client.go]: Not sure if this is the right direction to go. In the future we probably have subscriptions for whisper and swarm as well. These will be prefix with a different identifier. Maybe this method can accept the namespace as an argument?

File: rpc/client.go [hunk 2]
```diff
@@ -0,0 +1,737 @@
...
+	u, err := url.Parse(rawurl)
+	if err != nil {
+		return nil, err
+	}
+	switch u.Scheme {
+	case "http", "https":
+		return DialHTTP(rawurl)
+	case "ws", "wss":
+		return DialWebsocket(ctx, rawurl, "")
+	case "":
```
Comment (bas-vk) [inline rpc/client.go]: Please add "ipc" to this case statement. It allows people to use something like `ipc:///tmp/ethereum.sock`. The `attach` subcommand also allows for this.

Comment (fjl) [inline rpc/client.go]: The "ipc:" prefix is already stripped in cmd/geth

Comment (fjl) [inline rpc/client.go]: I'll figure out how to add this anyway.

File: rpc/client.go [hunk 3]
```diff
@@ -0,0 +1,737 @@
...
+	select {
+	case c.close <- struct{}{}:
+		<-c.didQuit
+	case <-c.didQuit:
+	}
+}
+
+// Call performs a JSON-RPC call with the given arguments
+// and unmarshals into result if no error occurred.
+func (c *Client) Call(result interface{}, method string, args ...interface{}) error {
```
Comment (bas-vk) [inline rpc/client.go]: I believe it is for client developers more logical to accept the method as first argument and the result as second. Also add a statement in the docs that result must be a pointer. The json unmarshal method states the following:
> Unmarshal parses the JSON-encoded data and stores the result in the value pointed to by result.

File: rpc/client.go [hunk 4]
```diff
@@ -0,0 +1,737 @@
...
+	return len(msg.ID) > 0 && msg.ID[0] != '{' && msg.ID[0] != '['
+}
+
+func (msg *jsonrpcMessage) String() string {
+	b, _ := json.Marshal(msg)
+	return string(b)
+}
+
+// Client represents a connection to an RPC server.
+type Client struct {
```
Comment (bas-vk) [inline rpc/client.go]: I know how you feel about this, but can we call this `JSONClient` instead of `Client`?

File: rpc/client.go [hunk 5]
```diff
@@ -0,0 +1,737 @@
...
+func newClientSubscription(c *Client, channel reflect.Value) *ClientSubscription {
+	sub := &ClientSubscription{
+		client:  c,
+		etype:   channel.Type().Elem(),
+		channel: channel,
+		// in is buffered so dispatch can continue even if the subscriber is slow.
+		in: make(chan json.RawMessage, clientSubscriptionBuffer),
+		// quit is buffered so dispatch can progress immediately while setting up the
+		// subscription in handleResponse.
+		quit: make(chan error),
```
Comment (fjl) [inline rpc/client.go]: This comment is outdated.

File: rpc/client.go [hunk 6]
```diff
@@ -0,0 +1,737 @@
...
+			glog.V(logger.Debug).Infof("<-readErr: %v", err)
+			c.closeRequestOps(err)
+			conn.Close()
+			reading = false
+
+		case conn = <-c.reconnected:
+			glog.V(logger.Debug).Infof("<-reconnected: (reading=%t) %v", reading, conn.RemoteAddr())
+			if reading {
+				// Wait for the previous read loop to exit. This is a rare case.
+				conn.Close()
```
Comment (fjl) [inline rpc/client.go]: This is broken.

File: rpc/client.go (hunk 4 above)
Comment (fjl) [inline rpc/client.go]: We decided against renaming in a call.

File: rpc/client.go (hunk 1 above)
Comment (fjl) [inline rpc/client.go]: We decided to deal with this later when there is an actual need for this.

File: rpc/client.go (hunk 2 above)
Comment (fjl) [inline rpc/client.go]: In the review call, Bas @bas-vk said that he doesn't feel very strongly about this.
I'll leave it up to the second reviewer to decide, but would rather not add support for
an ipc:// prefix.

File: rpc/client.go (hunk 3 above)
Comment (fjl) [inline rpc/client.go]: I'm still unsure about this. Would be nice to have a third opinion.

**Comment by fjl** [comment]
I'm still unsure about the parameter order for Call. At the moment the result is first.
@bas-vk said that this is weird and the method should be first. I tried changing to this but then
it's even weirder. It cannot be the last parameter because we use varargs for the actual arguments.

Our options are:

```go
var result int

// 1 (@fjl prefers this)
client.Call(&result, "service_method", "arg1", "arg2", "arg3")
client.CallContext(ctx, &result, "service_method", "arg1", "arg2", "arg3")

// 2 (@bas-vk prefers this)
client.Call("service_method", &result, "arg1", "arg2", "arg3")
client.CallContext(ctx, "service_method", &result, "arg1", "arg2", "arg3")

// 3 (nobody wants this)
client.Call("service_method", []interface{}{"arg1", "arg2", "arg3"}, &result)
client.CallContext(ctx, "service_method", []interface{}{"arg1", "arg2", "arg3"}, &result)
```

Would be nice to get a third opinion on this.

**Comment by fjl** [comment]
I've pushed an alternative implementation for subscriptions which is in line with the idea in https://github.com/ethereum/go-ethereum/pull/2798#issuecomment-232444816. 
The subscription now has an error channel, eliminating the need to close the caller-supplied event channel.

**Comment by fjl** [comment]
@karalabe @bas-vk PTAL. I have fixed the issues that came up during initial review.

File: rpc/client_example_test.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+// subscribeBlocks runs in its own goroutine and maintains
+// a subscription for new blocks.
+func subscribeBlocks(client *rpc.Client, subch chan Block) {
+	for i := 0; ; i++ {
+		if i > 0 {
+			time.Sleep(2 * time.Second)
+		}
+
+		// Subscribe to new blocks.
+		sub, err := client.EthSubscribe(subch, "newBlocks")
```
Comment (bas-vk) [inline rpc/client_example_test.go]: `newBlocks` accepts an argument that allows the client to control what is returned.
See https://github.com/ethereum/go-ethereum/blob/develop/internal/ethapi/api.go#L475

Update code and comment on line 33.

**Comment by bas-vk** [comment]
Code LGTM 👍 

We should think about the location of ethapi in the internal package. Large parts of the eth namespace are currently not usable from the outside. It would be nice if clients can use the argument and response types from the `ethapi/*` in their code.

File: accounts/abi/bind/backends/remote.go [hunk 8]
```diff
@@ -37,183 +33,78 @@ var _ bind.ContractBackend = (*rpcBackend)(nil)
...
 		return false, err
 	}
-	// Convert the response back to a Go byte slice and return
 	return len(common.FromHex(hex)) > 0, nil
 }
 
 // ContractCall implements ContractCaller.ContractCall, delegating the execution of
 // a contract call to the remote node, returning the reply to for local processing.
 func (b *rpcBackend) ContractCall(ctx context.Context, contract common.Address, data []byte, pending bool) ([]byte, error) {
-	// Pack up the request into an RPC argument
```
Comment (karalabe) [inline accounts/abi/bind/backends/remote.go]: Remember when you told me that changing comments just for the sake of style is not nice? ;)

**Comment by karalabe** [comment]
Code looks good to me per the previous review session. Since our RPC in general isn't tester the best and as the change is massive I can imagine a few corner cases might creep in, but we'll have to live with it (develop for a reason). I'd say merge it (after a squash). Please also rebase it on latest develop (I just want to see the circleci/hive tests green).

File: accounts/abi/bind/backends/remote.go (hunk 8 above)
Comment (fjl) [inline accounts/abi/bind/backends/remote.go]: Sorry

**Comment by fjl** [comment]
> We should think about the location of ethapi in the internal package. Large parts of the eth namespace are currently not usable from the outside. It would be nice if clients can use the argument and response types from the ethapi/* in their code.

I would like to wrap the 'raw client' with a typed client package (ethclient) next. My plan is to ensure that the argument types are neutral (e.g. types.Block, common.Hash, int) because the client's API should dual the in-process Go API.

**Comment by karalabe** [comment]
LGTM :+1: Squash and feel free to merge.

**Comment by fjl** [comment]
squashed
